stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -77,35 +73,88 @@
+;; get the aggregate public key for the given reward cycle (or none)
+(define-read-only (get-approved-aggregate-key (reward-cycle uint))
+    (map-get? aggregate-public-keys reward-cycle)
+)
+
 (define-private (is-in-voting-window (height uint) (reward-cycle uint))
-    (let ((last-cycle (unwrap-panic (contract-call? .signers stackerdb-get-last-set-cycle))))
+    (let ((last-cycle (unwrap-panic (contract-call? .signers get-last-set-cycle))))
         (and (is-eq last-cycle reward-cycle)
             (is-in-prepare-phase height))))
 
+(define-private (sum-weights (signer { signer: principal, weight: uint }) (acc uint))
+    (+ acc (get weight signer))
+)
+
+(define-private (get-total-weight (reward-cycle uint))
+    (match (map-get? cycle-total-weight reward-cycle)
+        total (ok total)
+        (let (
+                (signers (unwrap! (contract-call? .signers get-signers reward-cycle) (err ERR_FAILED_TO_RETRIEVE_SIGNERS)))
+                (total (fold sum-weights signers u0))
+            )
+            (map-set cycle-total-weight reward-cycle total)
+            (ok total)
+        )
+    )
+)
+
+;; Signer vote for the aggregate public key of the next reward cycle
+;;  The vote happens in the prepare phase of the current reward cycle but may be ran more than
+;;  once resulting in different 'rounds.' Each signer vote is based on the weight of stacked
+;;  stx tokens & fetched from the .signers contract. The vote is ran until the consensus 
+;;  threshold of 70% for a specific aggregate public key is reached. 
 (define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint))
     (let ((reward-cycle (+ u1 (burn-height-to-reward-cycle burn-block-height)))
             (tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
-            ;; one slot, one vote
-            (num-slots (try! (get-current-signer-slots signer-index)))
-            (new-total (+ num-slots (default-to u0 (map-get? tally tally-key)))))
-        (asserts! (is-in-voting-window burn-block-height reward-cycle) err-out-of-voting-window)
-        (asserts! (>= round (default-to u0 (map-get? rounds reward-cycle))) err-old-round)
-        (asserts! (is-eq (len key) u33) err-ill-formed-aggregate-public-key)
-        (asserts! (is-valid-aggregated-public-key key {reward-cycle: reward-cycle, round: round}) err-duplicate-aggregate-public-key)
-        (asserts! (map-insert votes {reward-cycle: reward-cycle, round: round, signer: tx-sender} {aggregate-public-key: key, reward-slots: num-slots}) err-duplicate-vote)
+            ;; vote by signer weight
+            (signer-weight (try! (get-current-signer-weight signer-index)))
+            (new-total (+ signer-weight (default-to u0 (map-get? tally tally-key))))
+            (total-weight (try! (get-total-weight reward-cycle))))
+        ;; Check we're in the prepare phase
+        (asserts! (is-in-voting-window burn-block-height reward-cycle) (err ERR_OUT_OF_VOTING_WINDOW))
```

Comment:
I am not sure we should have a limitation on voting window. what happens if the signers fail during the prepare phase to complete dkg...they should still be able ot perform dkg in the following reward cycle and the old signers continue to prop up the system until this actually happens. 

This makes me think I need to change my own signer logic to not immediately shut down a signer if it passes its own reward cycle...

Reply:
@setzeus updated this check in the latest.